features/exploredash/src/main/java/org/dash/wallet/features/exploredash/ui/dashdirect/dialogs/PurchaseGiftCardConfirmDialog.kt

```diff
@@ -190,69 +186,70 @@ class PurchaseGiftCardConfirmDialog : OffsetDialogFragment() {
+            R.drawable.ic_info_red,
+            getString(R.string.gift_card_purchase_failed),
+            getString(R.string.gift_card_error),
+            getString(R.string.cancel),
+            getString(R.string.try_again)
+        ).show(requireActivity()) { action?.invoke(it) }
     }
 
-    private suspend fun getGift(
+    private suspend fun getGiftCard(
         giftCardId: Long,
         transaction: Transaction,
-        merchant: Merchant?,
+        merchant: Merchant,
         paymentValue: Pair<Coin, Fiat>?
     ) {
         when (val response = viewModel.getGiftCardDetails(giftCardId)) {
             is ResponseResource.Success -> {
-                if (response.value?.successful == true) {
-                    showGiftCardDetailsDialog(merchant, paymentValue, transaction.txId, response?.value?.data)
+                if (response.value?.successful == true && response.value?.data != null) {
+                    showGiftCardDetailsDialog(merchant, paymentValue, transaction.txId, response.value!!.data!!)
                 }
             }
-            else -> {
+            is ResponseResource.Failure -> {
                 hideLoading()
-                Log.e(this::class.java.simpleName, "purchaseGiftCard error")
+                log.error("getGiftCardDetails error ${response.errorCode}: ${response.errorBody}")
                 showErrorRetryDialog {
                     if (it == true) {
                         showLoading()
-                        lifecycleScope.launch { getGift(giftCardId, transaction, merchant, paymentValue) }
+                        lifecycleScope.launch { getGiftCard(giftCardId, transaction, merchant, paymentValue) }
                     }
                 }
             }
+            else -> { }
         }
     }
     private fun showGiftCardDetailsDialog(
-        merchant: Merchant?,
+        merchant: Merchant,
         paymentValue: Pair<Coin, Fiat>?,
         txId: Sha256Hash,
-        data: GetGiftCardResponse.Data?
+        data: GetGiftCardResponse.Data
     ) {
-        GiftCardDetailsDialog.newInstance(
-                GiftCardDetailsDialogModel(
-                    merchantName = merchant?.name,
-                    merchantLogo = merchant?.logoLocation,
-                    giftCardPrice = GenericUtils.fiatToString(paymentValue?.second),
-                    transactionId = txId.toString(),
-                    giftCardNumber = data?.cardNumber,
-                    giftCardPin = data?.cardPin,
-                    barcodeImg = data?.barcodeUrl,
-                    giftCardCheckCurrentBalanceUrl = merchant?.website
-                )
-            )
-            .show(requireActivity())
-            .also {
-                val navController = findNavController()
-                navController.popBackStack(navController.graph.startDestinationId, false)
+        val giftCardId = "${ServiceName.DashDirect.lowercase()}+${data.id ?: -1}" // e.g. dashdirect+1234
```

Comment:
Even though we only have Dash Direct now, I think we need to have unique IDs across different possible gift card providers. `service+id` is the best I could come up with for now.